gdnative-core/src/core_types/string.rs

```diff
@@ -260,6 +260,68 @@ impl std::hash::Hash for GodotString {
     }
 }
 
+impl Add<&GodotString> for &GodotString {
+    type Output = GodotString;
+    #[inline]
+    fn add(self, other: &GodotString) -> GodotString {
+        GodotString::from_sys(unsafe { (get_api().godot_string_operator_plus)(&self.0, &other.0) })
+    }
+}
+
+impl<S> Add<S> for &GodotString 
+where 
+    S: AsRef<str>,
+{
+    type Output = GodotString;
+    #[inline]
+    fn add(self, other: S) -> GodotString {
+        self.add(&GodotString::from_str(other))
+    }
+}
+
+impl AddAssign<&GodotString> for GodotString {
+    #[inline]
+    fn add_assign(&mut self, other: &Self) {
+        *self = &*self + other;
+    }
+}
+
+impl<S> AddAssign<S> for GodotString
+where
+    S: AsRef<str>,
+{
+    #[inline]
+    fn add_assign(&mut self, other: S) {
+        self.add_assign(&GodotString::from_str(other))
+    }
+}
+
+impl PartialOrd for GodotString {
+    #[inline]
+    fn partial_cmp(&self, other: &GodotString) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for GodotString {
+    fn cmp(&self, other: &GodotString) -> Ordering {
+        if self == other {
+            Ordering::Equal
+        } else if unsafe { (get_api().godot_string_operator_less)(&self.0, &other.0) } {
+            Ordering::Less
+        } else {
+            Ordering::Greater
+        }
+    }
+}
+
+impl Index<usize> for GodotString {
+    type Output = u16;
```

Comment:
I've just realized that the size of `wchar_t` can change depending on the platform. On the build machine for docs.rs, for example, [it's `i32` instead](https://docs.rs/libc/0.2.77/libc/type.wchar_t.html). Perhaps it's not that great an idea to expose this in the high-level API after all, or maybe we can make a wrapper type for it that expose only a few high-level operations? e.g.

```rust
#[repr(transparent)] // so *const libc::wchar_t as *const GodotChar is valid
pub struct GodotChar(libc::wchar_t);

impl From<GodotChar> for char { /* .. */ }
impl PartialEq<char> for GodotChar { /* .. */ }
// ... and maybe some more operations?

impl Index<usize> for GodotString {
    type Output = GodotChar;
    fn index(&self, index: usize) -> &GodotChar { /* .. */ }
}
```

Reply:
Slightly different question: how performant is index access? it seems to be rather involved for the indirect function calls to C++. It may especially make a difference once people iterate over the string and access each character.

I also think the presence of `Index` may raise questions about `IndexMut` and `Iterator` traits.

Maybe instead of encouraging lots of operations on the original `GodotString`, we should advertise common idioms for easily converting the string to Rust and then perform string operations natively.

Reply:
`GodotString` does implement `Display` and `From<S>` for any `S: AsRef<str>`, but I agree that adding a `From<GodotString>` for `String` can be useful if it isn't available somewhere else already.

Reply:
If it implements `Display`, then the `GodotString::to_string()` implementation comes for free. An additional `From<GodotString>` might however be helpful for APIs that take `string: impl Into<String>` parameters. A trait `AsRef<str>` is probably not possible, since GodotString's internal representation is not a `str` slice (or `char` array)?

It's a general question whether you want `GodotString` to be a full-fledged string implementation, which might be a never-ending task, or rather just the "transport" type for APIs and promote standard `String` to do actual work. I would guess that copying typical strings (which are not huge) between two types can be faster than having indirect calls into the Godot engine for every string operation.

Reply:
Godot strings are not UTF-8, so dereferencing to `str` is impossible. Converting to and from `String` will re-encode the string which may or may not be worth it depending on the use case.

FYI, the bindings implement `FromVariant` and `ToVariant` for `String`, which means that you can simply take and return Rust `String`s in your exported methods, with the conversion automatically taken care of. Since 0.9, generated API methods are also generic over types that can be converted to `GodotString`, so you can also just pass `String`s and `&str`s into API methods, without having to care about the conversion. By all intents and purposes, the standard `String` *already is* a first-class string type, and users already don't have to learn a new API just to use the bindings. For those cases where a user specifically wants to use what's exposed over the GDNative API, I can't see why providing the option can be bad to the library.

Reply:
Ah, nice! Didn't know string support was already on a level where you could implement the bindings entirely without `GodotString`! 👍 

Yes, if it's clear that people have this option, then making `GodotString` more powerful makes sense. I just didn't see it in the docs/intro -- maybe we should add a paragraph to the `GodotString` doc, to highlight that there are 2 options of working with strings and each has their own trade-offs?

Furthermore, for intuitive APIs, it's probably good to add some operations in groups. E.g.:
* if `godotString + &string` works, so should `&string + godotString` and `godotString += &string`
* if you can get character at index `i`, you might want to write at that index as well (doesn't have to be `IndexMut`)
* ...

Reply:
>  I just didn't see it in the docs/intro ...

I didn't think "support for the language standard `String` type" was really worth a prominent position in the docs because... that should just be how it's expected to work? I thought a new user would declare the dependency, try using a string literal, see that it Just Works(tm), and get over it. It's also in the examples everywhere, like [this](https://github.com/godot-rust/godot-rust/blob/cd59838b59a952d059b2f301164802e60a046780/examples/scene_create/src/lib.rs#L137).

> maybe we should add a paragraph to the GodotString doc, to highlight that there are 2 options of working with strings and each has their own trade-offs?

More detailed documentation about the tradeoffs on the `GodotString` type would of course be helpful. A PR to add that would be welcome.

Reply:
My way of thinking was more "they have a custom string type, so I assume I'm supposed to work with that" 😉 

Here's the PR: https://github.com/godot-rust/godot-rust/pull/625